### DISCUSSION TITLE
[question] What's the difference between searx and searxng?

### DISCUSSION THREAD

**Discussion by contrun** [searxng/searxng#972] [Q&A] [answered]
Hi, I am new to both searx and searxng. I was trying to host my own instance of searx on raspberry pi. But searx does not provide docker image on arm64. So I found out this repo.

It seems searx is actively maintained and searx started from https://github.com/return42/searx. I presume @return42 is the main author of searxng. Judging from the commits history, @return42 contributed a lot to searx. I am curious about why there is a new fork. What additional features does searxng provide? I didn't find an answer in the documentation. I think this question merits an entry in the FAQ. What do you say?

**Comment by unixfox** [comment]
First asciimoo is the author of SearX.

Second dalf was the developer that maintained most of the code for a long time on SearX. He wanted to introduce new interesting features into the core of Searx but asciimoo didn't share the same vision as dalf about that so dalf left the SearX project and started his own fork.

In contrary to SearX, SearXNG is going to evolve a lot during the coming months now that dalf is free to introduce any changes he wants. Maybe @dalf can provide what kind of new features SearXNG is going to have compared to the original SearX.

**Comment by dalf** [comment]
@contrun

Currently:

---

The docker image supports amd64, arm64, arm/v7 : this can be backported to searx/searx.


---

The HTTP client has changed: [requests](https://github.com/psf/requests) has been replaced by [httpx](https://github.com/encode/httpx).

httpx supports HTTP/2 (can be disabled / enable per engine).

If you specify multiple proxies or local IP addresses, searxng will make rotation per engine.

There is are two new options `retries`  and `retry_on_http_error ` to automatically retries a failed HTTP request ( see https://searxng.github.io/searxng/admin/settings.html ).

---

When an engine can't return a result the error message is a link:
![image](https://user-images.githubusercontent.com/1594191/116808153-91a51c00-ab37-11eb-8242-6d8ff24b80d9.png)

This link goes to this page:
![image](https://user-images.githubusercontent.com/1594191/116808176-b3060800-ab37-11eb-9d9f-b9498e372d5f.png)

The percentage is actually a ratio from 0 to 20 multiply by 5 (to hide the number of requests).

The "Submit a new issue on Github" will start a new issue with the error logs : this process may change in the future to avoid duplicate issues.

---

If an engine is suspended, it is reported:
![image](https://user-images.githubusercontent.com/1594191/116808274-37588b00-ab38-11eb-9d3a-214009e0d9f1.png)

In searx/searx, you won't see any result / any error. This feature most probably can be backported to searx/searx.

---

In the preferences, the right part has changed:
![image](https://user-images.githubusercontent.com/1594191/116808403-de3d2700-ab38-11eb-999a-b1c8612a3176.png)

![image](https://user-images.githubusercontent.com/1594191/116808260-1c861680-ab38-11eb-9eeb-c0cfeb4fde6b.png)

* the response time displays the median time, P80, P95 (instead of the average in searx/searx)
* the reliability column displays the percentage of successful user request:
    * no crash
    * if the ['simple' checker test](https://github.com/searxng/searxng/blob/2756e170ce9f378362002d916f09649c350f170e/searx/search/processors/online.py#L169-L172) fail the reliability is set to 0 (the simple checker test checks if some simple queries returns at least one result)
* in the above example, there is a exclamation mark for the gentoo engine in the "Supported language" column: 
    * the engine declares to support the language selection
    * but the checker test has failed.

---

The `/stats` page has been updated:
![image](https://user-images.githubusercontent.com/1594191/116808638-345e9a00-ab3a-11eb-8909-c7d1ef68d2e4.png)

---

Notes: 
* The checker integration needs some adjustments.
* These additional features require to store [more data in searx server memory](https://github.com/searxng/searxng/blob/2756e170ce9f378362002d916f09649c350f170e/searx/metrics/__init__.py#L64-L98) (for comparison searx/searx stores [the total number of request and the total time](https://github.com/searx/searx/blob/3a0f896b68b24957421c593dc2bfacdd36b8bcb3/searx/engines/__init__.py#L143-L151).
* The checker when enable sends request automatically which may trigger the search engines (same few queries per engine but every 24h as [it is configured now](https://github.com/searxng/searxng/blob/2756e170ce9f378362002d916f09649c350f170e/searx/settings.yml#L116-L118);, it may change in the future).

**Comment by return42** [comment]
> It seems searx is actively maintained and searx started from https://github.com/return42/searx. I presume @return42 is the main author of searxng. 

No. SearXNG is a fork from https://github.com/searx/searx and @dalf is the founder of this repository.

> Judging from the commits history, @return42 contributed a lot to searx. I am curious about why there is a new fork. What additional features does searxng provide? I didn't find an answer in the documentation. I think this question merits an entry in the FAQ. What do you say?

In deed, I think it is worth to mentioned. SearXNG is maintained only for the 10 days and we applied 45 PRs .. there is a lot of more work to differentiate the fork .. this may take a little longer, please be patient with us :-)

**Comment by kvch** [comment]
@contrun I am a maintainer for the original searx repository. I am actively cherry-picking commits from this fork. For example, I am preparing for removing requests from searx in favour of httpx (https://github.com/searx/searx/pull/2800). But also, you can see in the recent commit history of the repo that changes are integrated into searx. I believe most of the changes from this repository will make it back to searx/searx.

The difference of vision between the maintainers is regarding how privacy-respecting searx should be. The original guidelines of searx state that features that risk the privacy of users are not welcomed to the repo. Consider this fork of searx as a shinier version with more features that might expose you to others. Searx is going to be kept minimal with fewer features and less danger to your privacy. Also, I do not think that searxng is going to abandon all privacy-protecting features, but that part should be clarified by @dalf.

It does not mean that searx is not maintained anymore. It just has a higher bar when it comes to privacy. If you would like to get privacy without compromising, I suggest you use searx. If you are willing to trade some of the protections searx provides for more features, you should choose this fork.

@return42 I agree we need a FAQ. I suggest we collaborate on a response to avoid future conflicts.

**Comment by return42** [comment]
> The difference of vision between the maintainers is regarding how privacy-respecting searx should be ... Consider this fork of searx as a shinier version with more features that might expose you to others.

Sorry, what you postulate here is nonsense. SearXNG's team has never announced any "visions" nor any change in the aspects of "privacy".

In practice: In the past I have send my contributions to searx and there have never been any complaints regarding "privacy" .. now I contribute to SearXNG and there has nothing changed about my opinions on "privacy".

If one comes to the conclusion "privacy-respecting" is violated in any manner, I would be pleased to receive his [issue](https://github.com/searxng/searxng/issues/new/choose).

Back to "visions", I do not have big "visions" about SearXNG's future / IMO open-source projects with to much visions often tend to fail.

My vision is, that both branches (searx & SearXNG) can benefit from each other, that is why I track searx and merge everything what is relevant into SearXNG (BTW: there are several other forks we can benefit from).

Another "vision" of mine is to keep the engines running with some quality .. the features @dalf described [above](https://github.com/searxng/searxng/issues/46#issuecomment-830780519) will help developers to boost the quality of engines.

The opener asked "what is the difference" .. for me it is very clear, the searx team dropped me from the maintainer list.  Here at SearXNG I fell more welcome and I see an accelerated progress in my work.

**Comment by asciimoo** [comment]
We had different opinions with dalf about features and how can those reduce the protection of searx regarding to information leaks. A good example is the enhanced stats/checker page (shown above) which can provide more information to a potential "attacker" in theory. These kind of differences lead to the fork in my opinion, so what is kvch saying isn't nonsense at all. We at searx do not want to reduce the privacy even if the possible privacy violation is theoretical. So we don't want to enhance the stats page, nor want to introduce default enabled checkers, more over we are thinking to disable it completely by default and making it optionally available only to provide better privacy protection to the instances. Especially because in the past 6-7 years we didn't really received requests to improve our statistics/diagnostics.

> The opener asked "what is the difference" .. for me it is very clear, the searx team dropped me from the maintainer list. Here at SearXNG I fell more welcome and I see an accelerated progress in my work.

We are always open to contributions, the issue there was more like communication differences and your harsh/categorical statements just like in your previous post saying "nonsense" to something where you don't have all the available information about a topic. This doesn't mean that we are not open to contributions we just felt that this kind of explicit communication style is far from our standards and from the image that we intent to show to our contributors/users. I think a good proof is that we merged a big PR from you in the following days in the searx repo and that's the same with all of your contributions ever since you are not part of the maintainer team.

**Comment by kvch** [comment]
> > The difference of vision between the maintainers is regarding how privacy-respecting searx should be ... Consider this fork of searx as a shinier version with more features that might expose you to others.
> 
> Sorry, what you postulate here is nonsense. SearXNG's team has never announced any "visions" nor any change in the aspects of "privacy".

What I stated is not based on public announcements of SearXNG. We had numerous internal discussions in the last 6-7 months about several topics in searx before this fork was created.

> In practice: In the past I have send my contributions to searx and there have never been any complaints regarding "privacy" .. now I contribute to SearXNG and there has nothing changed about my opinions on "privacy".

I never said and I do not believe that your views on privacy has changed.

> Back to "visions", I do not have big "visions" about SearXNG's future / IMO open-source projects with to much visions often tend to fail.

You are probably right. :)

> My vision is, that both branches (searx & SearXNG) can benefit from each other, that is why I track searx and merge everything what is relevant into SearXNG (BTW: there are several other forks we can benefit from).

I agree. Also, I believe that both searx and SearXNG has their target audiences. We can live alongside each other without problems.

> Another "vision" of mine is to keep the engines running with some quality .. the features @dalf described [above](https://github.com/searxng/searxng/issues/46#issuecomment-830780519) will help developers to boost the quality of engines.

The features are indeed valuable. My problem is that it is going to be enabled by default on all SearXNG instances. In my opinion these features are useful tools for **developers** of searx. I suggested it to be a separate development tool, however, my concerns were dismissed. I still believe that the software should serve the users. We should not enable features e.g. metrics that benefits the software by giving better diagnostics, but those metrics of users will be exposed to administrators. I am willing to trade in a few unnoticed broken engines for better privacy of our users.

Also, I respected the decision to add more advanced metrics were to searx as they were disabled by default. At least we provide privacy by default. If an administrator wants to enable it, they can do it. I consider it an acceptable compromise.

To reiterate I am not here to generate conflict. I think both searx and SearXNG have merit with slightly different things to offer. I would like to keep the good relationship between the two projects and maintainer groups.

> The opener asked "what is the difference" .. for me it is very clear, the searx team dropped me from the maintainer list. Here at SearXNG I fell more welcome and I see an accelerated progress in my work.

Your contributions when proven valuable were always accepted. I am sorry if I ever made you feel not welcomed. It was never my intention.

**Comment by return42** [comment]
@asciimoo 

> We are always open to contributions, the issue there was more like communication differences and your harsh/categorical statements just like in your previous post saying "nonsense" to something where you don't have all the available information about a topic. 

My tone is sometimes harsh, with newcomers I'm always patient but experienced users should think before purport ..

> We had different opinions 

That exactly is the point, we have different opinions, but to postulate ..

> Consider this fork of searx as a shinier version with more features that might expose you to others.

is nonsense. 

> A good example is the enhanced stats/checker page 

IMO a better example for privacy violation are external bangs. At the end: the real risks are not external bangs nor stats, the real risk is how a searx(ng) instance is hosted and maintained.

@kvch 

> To reiterate I am not here to generate conflict. I think both searx and SearXNG have merit with slightly different things to offer. I would like to keep the good relationship between the two projects and maintainer groups.

I'm with you, lets look forward.  It was only your remark about privacy what rumbles me and I wanted to clarify. Excuse if my tone was harsh to you ..

**Comment by dit7ya** [comment]
Till proper FAQs are written, it would be very helpful if there is a link to this thread in the project Readme. I can add a PR for that.

**Comment by dalf** [comment]
@dit7ya Still not a FAQ but a detail change log may help: https://github.com/searxng/searxng/wiki/Changes-from-version-1.0.0

Also I have opened https://github.com/searxng/searxng/pull/283

Don't hesitate to suggest a better wording

**Comment by return42** [comment, marked as answer]
Two years later we can say, SearXNG is the active project.
